File: app/src/main/kotlin/com/igorganapolsky/vibratingwatchapp/presentation/settings/SetTimerActivity.java
package com.igorganapolsky.vibratingwatchapp.presentation.settings;

import androidx.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import com.google.android.material.tabs.TabLayout;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatImageView;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;
import com.igorganapolsky.vibratingwatchapp.R;
import com.igorganapolsky.vibratingwatchapp.util.StepActionListener;
import com.igorganapolsky.vibratingwatchapp.util.SwipeRestrictViewPager;
import com.igorganapolsky.vibratingwatchapp.presentation.ViewModelFactory;
import com.igorganapolsky.vibratingwatchapp.domain.repo.TimerEntity;
import com.igorganapolsky.vibratingwatchapp.domain.model.TimerModel;
import com.igorganapolsky.vibratingwatchapp.domain.model.Time;

import java.util.Locale;

import static com.igorganapolsky.vibratingwatchapp.domain.repo.TimerEntity.TIMER_ID;

public class SetTimerActivity extends AppCompatActivity implements View.OnClickListener, StepActionListener {

    public static final int SETTING_SUCCESS_CODE = 101;

    private SetTimerViewModel mViewModel;

    private SwipeRestrictViewPager vpWizard;

    private TextView tvTimeHours;
    private TextView tvTimeMinutes;
    private TextView tvTimeSeconds;
    private boolean isProgressChanging = false;
    private boolean isSwipeGranted = false;

    public static Intent createIntent(Context context, int timerId) {
        Intent intent = new Intent(context, SetTimerActivity.class);
        intent.putExtra(TimerEntity.TIMER_ID, timerId);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_set_timer);
        mViewModel = ViewModelProviders.of(this, ViewModelFactory.getInstance()).get(SetTimerViewModel.class);

        setupViewModel();
        setupView();
        setupObservers();
    }

    private void setupViewModel() {
        Bundle bundle = getIntent().getExtras();
        int currentId = bundle != null ? bundle.getInt(TIMER_ID) : TimerModel.UNDEFINE_ID;
        mViewModel.setCurrentModelId(currentId);
    }

    private void setupView() {
        vpWizard = findViewById(R.id.vpWizard);
        AppCompatImageView ivNextPage = findViewById(R.id.ivNextPage);

        tvTimeHours = findViewById(R.id.tvTimeHours);
        tvTimeMinutes = findViewById(R.id.tvTimeMinutes);
        tvTimeSeconds = findViewById(R.id.tvTimeSeconds);

        TabLayout tlDots = findViewById(R.id.tlDots);

        vpWizard.setAdapter(new SetTimerPageAdapter(getSupportFragmentManager()));
        vpWizard.setOffscreenPageLimit(2);
        ivNextPage.setOnClickListener(this);

        tlDots.setupWithViewPager(vpWizard, true);
        disableTabs(tlDots);
    }

    private void setupObservers() {
        mViewModel.getSetupData().observe(this, this::updateTime);
        mViewModel.getSwipeState().observe(this, this::updateSwipeStateIfNeeded);
        mViewModel.getTimerData().observe(this, (timerModel -> {
            if (timerModel == null) return;
            updateTimerData(Time.HOURS, timerModel.getHours());
            updateTimerData(Time.MINUTES, timerModel.getMinutes());
            updateTimerData(Time.SECONDS, timerModel.getSeconds());
        }));
    }

    private void updateTime(Time setup) {
        updateTimerData(setup, mViewModel.getCurrentTimeValue());
    }


    private void updateTimerData(Time timerSetup, int newValue) {
        switch (timerSetup) {
            case HOURS:
                tvTimeHours.setText(String.format(Locale.ENGLISH, "%02d", newValue));
                break;
            case MINUTES:
                tvTimeMinutes.setText(String.format(Locale.ENGLISH, "%02d", newValue));
                break;
            case SECONDS:
                tvTimeSeconds.setText(String.format(Locale.ENGLISH, "%02d", newValue));
                break;
        }
    }

    private void updateSwipeStateIfNeeded(Boolean isSwipeGranted) {
        if (isProgressChanging) {
            this.isSwipeGranted = isSwipeGranted;
        } else {
            vpWizard.setIsSwipeAvailable(isSwipeGranted);
        }
    }

    @Override
    public void onClick(View view) {
        if (!isProgressChanging && vpWizard.isSwipeAvailable()) {
            int currentPage = vpWizard.getCurrentItem();
            if (currentPage < 2) {
                vpWizard.setCurrentItem(currentPage + 1);
            } else {
                mViewModel.saveTimer();
                setResult(SETTING_SUCCESS_CODE);
                finish();
            }
        }
    }

    @Override
    public void onActionStart() {
        isProgressChanging = true;
        vpWizard.setIsSwipeAvailable(false);
    }

    @Override
    public void onActionEnd() {
        isProgressChanging = false;
        vpWizard.setIsSwipeAvailable(isSwipeGranted);
    }

    private void disableTabs(TabLayout layout) {
        LinearLayout tabStrip = ((LinearLayout) layout.getChildAt(0));
        for (int i = 0; i < tabStrip.getChildCount(); i++) {
            tabStrip.getChildAt(i).setOnTouchListener((v, event) -> true);
        }
    }
}

File: app/src/main/kotlin/com/igorganapolsky/vibratingwatchapp/presentation/settings_screen/SetTimerActivity.kt
package com.igorganapolsky.vibratingwatchapp.presentation.settings_screen

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatImageView
import com.google.android.material.tabs.TabLayout
import com.igorganapolsky.vibratingwatchapp.core.IStepActionListener
import com.igorganapolsky.vibratingwatchapp.core.SwipeRestrictViewPager
import com.igorganapolsky.vibratingwatchapp.core.extensions.observe
import com.igorganapolsky.vibratingwatchapp.data.TimerEntity.TIMER_ID
import com.igorganapolsky.vibratingwatchapp.data.TimeMeasurement
import com.igorganapolsky.vibratingwatchapp.domain.model.TimerModel
import com.igorganapolsky.vibratingwatchapp.presentation.home_screen.TimerViewModel
import java.util.*

/**
 * Screen responsible for creating a timer.
 */
class SetTimerActivity : AppCompatActivity(), View.OnClickListener, IStepActionListener {
    //    private var mViewModel: TimerViewModel? = null
    private var vpWizard: SwipeRestrictViewPager? = null
    private var tvTimeHours: TextView? = null
    private var tvTimeMinutes: TextView? = null
    private var tvTimeSeconds: TextView? = null
    private var isProgressChanging = false
    private var isSwipeGranted = false

    private val mViewModel by viewModels<TimerViewModel>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(com.igorganapolsky.vibratingwatchapp.R.layout.activity_set_timer)
//        mViewModel = ViewModelProviders.of(this, ViewModelFactory.instance)
//            .get(TimerViewModel::class.java)

        setupViewModel()
        setupView()
        setupObservers()
    }

    private fun setupViewModel() {
        val bundle = intent.extras
        val currentId = bundle?.getInt(TIMER_ID) ?: TimerModel.UNDEFINE_ID
        mViewModel.setCurrentModelId(currentId)
    }

    private fun setupView() {
        vpWizard = findViewById(com.igorganapolsky.vibratingwatchapp.R.id.vpWizard)
        val ivNextPage =
            findViewById<AppCompatImageView>(com.igorganapolsky.vibratingwatchapp.R.id.ivNextPage)

        tvTimeHours = findViewById(com.igorganapolsky.vibratingwatchapp.R.id.tvTimeHours)
        tvTimeMinutes = findViewById(com.igorganapolsky.vibratingwatchapp.R.id.tvTimeMinutes)
        tvTimeSeconds = findViewById(com.igorganapolsky.vibratingwatchapp.R.id.tvTimeSeconds)

        val tlDots = findViewById<TabLayout>(com.igorganapolsky.vibratingwatchapp.R.id.tlDots)

        vpWizard!!.adapter = SetTimerPageAdapter(supportFragmentManager)
        vpWizard!!.offscreenPageLimit = 2
        ivNextPage.setOnClickListener(this)

        tlDots.setupWithViewPager(vpWizard, true)
        disableTabs(tlDots)
    }

    private fun setupObservers() {
        mViewModel.getCountdownData().observe(this, ::updateTime)
        mViewModel.getSwipeState().observe(this, ::updateSwipeStateIfNeeded)
        mViewModel.getTimerData().observe(this) { timerModel ->
            updateTimerData(TimeMeasurement.HOURS, timerModel.hours)
            updateTimerData(TimeMeasurement.MINUTES, timerModel.minutes)
            updateTimerData(TimeMeasurement.SECONDS, timerModel.seconds)
        }
    }

    private fun updateTime(setupTimeMeasurement: TimeMeasurement) {
        updateTimerData(setupTimeMeasurement, mViewModel.currentTimeValue)
    }


    private fun updateTimerData(timerSetup: TimeMeasurement, newValue: Int) {
        when (timerSetup) {
            TimeMeasurement.HOURS -> tvTimeHours!!.text =
                String.format(Locale.ENGLISH, "%02d", newValue)
            TimeMeasurement.MINUTES -> tvTimeMinutes!!.text =
                String.format(Locale.ENGLISH, "%02d", newValue)
            TimeMeasurement.SECONDS -> tvTimeSeconds!!.text =
                String.format(Locale.ENGLISH, "%02d", newValue)
        }
    }

    private fun updateSwipeStateIfNeeded(isSwipeGranted: Boolean?) {
        if (isProgressChanging) {
            this.isSwipeGranted = isSwipeGranted!!
        } else {
            vpWizard!!.isSwipeAvailable = isSwipeGranted!!
        }
    }

    override fun onClick(view: View) {
        if (!isProgressChanging && vpWizard!!.isSwipeAvailable) {
            val currentPage = vpWizard!!.currentItem
            if (currentPage < 2) {
                vpWizard!!.currentItem = currentPage + 1
            } else {
                mViewModel.saveTimer()
                setResult(SETTING_SUCCESS_CODE)
                finish()
            }
        }
    }

    override fun onActionStart() {
        isProgressChanging = true
        vpWizard!!.isSwipeAvailable = false
    }

    override fun onActionEnd() {
        isProgressChanging = false
        vpWizard!!.isSwipeAvailable = isSwipeGranted
    }

    private fun disableTabs(layout: TabLayout) {
        val tabStrip = layout.getChildAt(0) as LinearLayout
        for (i in 0 until tabStrip.childCount) {
            tabStrip.getChildAt(i).setOnTouchListener { v, event -> true }
        }
    }

    companion object {
        const val SETTING_SUCCESS_CODE = 101

        fun createIntent(context: Context, timerId: Int): Intent {
            val intent = Intent(context, SetTimerActivity::class.java)
            intent.putExtra(TIMER_ID, timerId)
            return intent
        }
    }
}
